File: strategies/base_strategy.py
from abc import ABC, abstractmethod
from datetime import datetime
from typing import Dict, Any, Optional

class BaseStrategy(ABC):
    """
    Abstract Base Class for Cinema Ticket Release Strategies.
    All site-specific or target-specific monitoring strategies must inherit from this class.
    """
    def __init__(self, strategy_id: str, name: str, description: str):
        self.strategy_id = strategy_id
        self.name = name
        self.description = description

    @abstractmethod
    def inspect(self, target: Dict[str, Any]) -> Dict[str, Any]:
        """
        Inspect target cinema page and return standardized check result dict:
        {
            "status": "AVAILABLE" | "COMING_SOON" | "UNAVAILABLE" | "ERROR",
            "is_available": bool,
            "movie_title": str,
            "booking_url": Optional[str],
            "details": str,
            "raw_match": Optional[Dict],
            "timestamp": str (ISO)
        }
        """
        pass

    def format_result(
        self,
        status: str,
        is_available: bool,
        movie_title: str,
        booking_url: Optional[str] = None,
        details: str = "",
        raw_match: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        return {
            "status": status,
            "is_available": bool(is_available),
            "movie_title": movie_title or "Unknown Movie",
            "booking_url": booking_url,
            "details": details or "",
            "raw_match": raw_match or {},
            "timestamp": datetime.utcnow().isoformat() + "Z"
        }

File: server/strategies/baseStrategy.js
/**
 * Base strategy interface for Cinema Ticket Monitors
 */
class BaseStrategy {
  constructor(id, name, description) {
    this.id = id;
    this.name = name;
    this.description = description;
  }

  /**
   * Inspect target website and determine ticket availability
   * @param {Object} target Target object configuration
   * @returns {Promise<Object>} Check result object
   */
  async inspect(target) {
    throw new Error(`inspect() method must be implemented by strategy ${this.id}`);
  }

  /**
   * Helper to format standard check result
   */
  formatResult({ status, isAvailable, movieTitle, bookingUrl, details, rawMatch }) {
    return {
      status, // 'AVAILABLE' | 'COMING_SOON' | 'UNAVAILABLE' | 'ERROR'
      isAvailable: Boolean(isAvailable),
      movieTitle: movieTitle || 'Unknown Movie',
      bookingUrl: bookingUrl || null,
      details: details || '',
      rawMatch: rawMatch || null,
      timestamp: new Date().toISOString()
    };
  }
}

module.exports = BaseStrategy;
